File: aac/api/typescript/src/output/TTSAdapter.ts
/**
 * WIA AAC TTS Adapter
 * Phase 4: WIA Ecosystem Integration
 *
 * Text-to-Speech adapter using Web Speech API
 */

import { BaseOutputAdapter } from './IOutputAdapter';
import {
  OutputOptions,
  OutputState,
  Voice,
  OutputError,
  OutputErrorCode
} from './types';

/**
 * TTS adapter interface
 */
export interface ITTSAdapter {
  /**
   * Get available voices
   */
  getVoices(): Promise<Voice[]>;

  /**
   * Set voice
   * @param voiceId - Voice ID
   */
  setVoice(voiceId: string): void;

  /**
   * Pause output
   */
  pause(): void;

  /**
   * Resume output
   */
  resume(): void;
}

/**
 * Web Speech API TTS Adapter
 */
export class WebSpeechTTSAdapter extends BaseOutputAdapter implements ITTSAdapter {
  readonly type = 'tts' as const;
  readonly name = 'WebSpeechTTS';

  private synth: SpeechSynthesis | null = null;
  private currentUtterance: SpeechSynthesisUtterance | null = null;
  private selectedVoice: SpeechSynthesisVoice | null = null;
  private defaultOptions: OutputOptions = {
    speed: 1.0,
    volume: 1.0
  };

  async initialize(options?: OutputOptions): Promise<void> {
    if (typeof window === 'undefined' || !window.speechSynthesis) {
      throw new OutputError(
        OutputErrorCode.NOT_AVAILABLE,
        'Web Speech API not available',
        false
      );
    }

    this.synth = window.speechSynthesis;
    if (options) {
      this.defaultOptions = { ...this.defaultOptions, ...options };
    }

    // Wait for voices to load
    await this.waitForVoices();

    this._state = 'idle';
    this.emit('start', { message: 'TTS adapter initialized' });
  }

  private async waitForVoices(): Promise<void> {
    return new Promise((resolve) => {
      const voices = this.synth!.getVoices();
      if (voices.length > 0) {
        resolve();
        return;
      }

      this.synth!.onvoiceschanged = () => {
        resolve();
      };

      // Timeout fallback
      setTimeout(resolve, 1000);
    });
  }

  async getVoices(): Promise<Voice[]> {
    if (!this.synth) {
      throw new OutputError(
        OutputErrorCode.ADAPTER_NOT_READY,
        'TTS adapter not initialized',
        true
      );
    }

    const voices = this.synth.getVoices();
    return voices.map(v => ({
      id: v.voiceURI,
      name: v.name,
      language: v.lang,
      local: v.localService
    }));
  }

  setVoice(voiceId: string): void {
    if (!this.synth) {
      throw new OutputError(
        OutputErrorCode.ADAPTER_NOT_READY,
        'TTS adapter not initialized',
        true
      );
    }

    const voices = this.synth.getVoices();
    const voice = voices.find(v => v.voiceURI === voiceId);
    if (voice) {
      this.selectedVoice = voice;
    }
  }

  async output(text: string, options?: OutputOptions): Promise<void> {
    if (!this.synth) {
      throw new OutputError(
        OutputErrorCode.ADAPTER_NOT_READY,
        'TTS adapter not initialized',
        true
      );
    }

    if (this._state === 'outputting') {
      this.stop();
    }

    const mergedOptions = { ...this.defaultOptions, ...options };

    return new Promise((resolve, reject) => {
      const utterance = new SpeechSynthesisUtterance(text);

      if (mergedOptions.language) {
        utterance.lang = mergedOptions.language;
      }
      if (mergedOptions.speed !== undefined) {
        utterance.rate = mergedOptions.speed;
      }
      if (mergedOptions.volume !== undefined) {
        utterance.volume = mergedOptions.volume;
      }
      if (this.selectedVoice) {
        utterance.voice = this.selectedVoice;
      }

      utterance.onstart = () => {
        this._state = 'outputting';
        this.emit('start', { text });
      };

      utterance.onend = () => {
        this._state = 'idle';
        this.emit('end', { text });
        resolve();
      };

      utterance.onerror = (event) => {
        this._state = 'error';
        const error = new OutputError(
          OutputErrorCode.OUTPUT_FAILED,
          `TTS error: ${event.error}`,
          true,
          event
        );
        this.emit('error', error);
        reject(error);
      };

      utterance.onpause = () => {
        this._state = 'paused';
        this.emit('pause');
      };

      utterance.onresume = () => {
        this._state = 'outputting';
        this.emit('resume');
      };

      this.currentUtterance = utterance;
      this.synth!.speak(utterance);
    });
  }

  stop(): void {
    if (this.synth) {
      this.synth.cancel();
      this._state = 'idle';
      this.currentUtterance = null;
    }
  }

  pause(): void {
    if (this.synth && this._state === 'outputting') {
      this.synth.pause();
      this._state = 'paused';
    }
  }

  resume(): void {
    if (this.synth && this._state === 'paused') {
      this.synth.resume();
      this._state = 'outputting';
    }
  }

  isAvailable(): boolean {
    return typeof window !== 'undefined' && !!window.speechSynthesis;
  }

  async dispose(): Promise<void> {
    this.stop();
    this.synth = null;
    this.selectedVoice = null;
    this.handlers.clear();
  }
}

/**
 * Mock TTS Adapter for testing
 */
export class MockTTSAdapter extends BaseOutputAdapter implements ITTSAdapter {
  readonly type = 'tts' as const;
  readonly name = 'MockTTS';

  private mockVoices: Voice[] = [
    { id: 'mock-ko', name: 'Mock Korean', language: 'ko-KR', gender: 'neutral', local: true },
    { id: 'mock-en', name: 'Mock English', language: 'en-US', gender: 'neutral', local: true }
  ];

  private selectedVoiceId: string | null = null;

  async initialize(): Promise<void> {
    this._state = 'idle';
    this.emit('start', { message: 'Mock TTS adapter initialized' });
  }

  async getVoices(): Promise<Voice[]> {
    return this.mockVoices;
  }

  setVoice(voiceId: string): void {
    this.selectedVoiceId = voiceId;
  }

  async output(text: string, options?: OutputOptions): Promise<void> {
    this._state = 'outputting';
    this.emit('start', { text });

    // Simulate TTS duration based on text length
    const duration = Math.max(text.length * 50, 500);
    const speed = options?.speed || 1.0;

    await new Promise(resolve => setTimeout(resolve, duration / speed));

    this._state = 'idle';
    this.emit('end', { text });
  }

  stop(): void {
    this._state = 'idle';
  }

  pause(): void {
    if (this._state === 'outputting') {
      this._state = 'paused';
      this.emit('pause');
    }
  }

  resume(): void {
    if (this._state === 'paused') {
      this._state = 'outputting';
      this.emit('resume');
    }
  }

  isAvailable(): boolean {
    return true;
  }

  async dispose(): Promise<void> {
    this.handlers.clear();
  }
}

File: aac/api/python/wia_aac/output/tts_adapter.py
"""
WIA AAC TTS Adapter
Phase 4: WIA Ecosystem Integration

Text-to-Speech adapter
"""

import asyncio
from abc import abstractmethod
from typing import Optional, List

from .base_output import BaseOutputAdapter
from .types import (
    OutputType,
    OutputState,
    OutputOptions,
    OutputEventType,
    Voice,
    OutputError,
    OutputErrorCode
)


class ITTSAdapter(BaseOutputAdapter):
    """TTS adapter interface"""

    @abstractmethod
    async def get_voices(self) -> List[Voice]:
        """Get available voices"""
        pass

    @abstractmethod
    def set_voice(self, voice_id: str) -> None:
        """Set voice"""
        pass

    @abstractmethod
    def pause(self) -> None:
        """Pause output"""
        pass

    @abstractmethod
    def resume(self) -> None:
        """Resume output"""
        pass


class MockTTSAdapter(ITTSAdapter):
    """Mock TTS Adapter for testing"""

    def __init__(self):
        super().__init__()
        self._mock_voices = [
            Voice(id="mock-ko", name="Mock Korean", language="ko-KR", gender="neutral", local=True),
            Voice(id="mock-en", name="Mock English", language="en-US", gender="neutral", local=True),
            Voice(id="mock-ja", name="Mock Japanese", language="ja-JP", gender="neutral", local=True)
        ]
        self._selected_voice_id: Optional[str] = None
        self._default_options = OutputOptions()
        self._is_paused = False

    @property
    def type(self) -> OutputType:
        return OutputType.TTS

    @property
    def name(self) -> str:
        return "MockTTS"

    async def initialize(self, options: Optional[OutputOptions] = None) -> None:
        """Initialize the adapter"""
        if options:
            self._default_options = options
        self._state = OutputState.IDLE
        self._emit(OutputEventType.START, {"message": "Mock TTS adapter initialized"})

    async def get_voices(self) -> List[Voice]:
        """Get available voices"""
        return self._mock_voices

    def set_voice(self, voice_id: str) -> None:
        """Set voice"""
        self._selected_voice_id = voice_id

    async def output(self, text: str, options: Optional[OutputOptions] = None) -> None:
        """Output text as TTS"""
        if self._state == OutputState.OUTPUTTING:
            self.stop()

        merged_options = OutputOptions(
            language=options.language if options else self._default_options.language,
            voice=options.voice if options else self._default_options.voice,
            speed=options.speed if options else self._default_options.speed,
            volume=options.volume if options else self._default_options.volume
        )

        self._state = OutputState.OUTPUTTING
        self._is_paused = False
        self._emit(OutputEventType.START, {"text": text})

        # Simulate TTS duration based on text length
        duration = max(len(text) * 0.05, 0.5)
        speed = merged_options.speed or 1.0

        try:
            await asyncio.sleep(duration / speed)

            if self._state == OutputState.OUTPUTTING:
                self._state = OutputState.IDLE
                self._emit(OutputEventType.END, {"text": text})
        except asyncio.CancelledError:
            self._state = OutputState.IDLE
            raise

    def stop(self) -> None:
        """Stop output"""
        self._state = OutputState.IDLE
        self._is_paused = False

    def pause(self) -> None:
        """Pause output"""
        if self._state == OutputState.OUTPUTTING:
            self._state = OutputState.PAUSED
            self._is_paused = True
            self._emit(OutputEventType.PAUSE)

    def resume(self) -> None:
        """Resume output"""
        if self._state == OutputState.PAUSED:
            self._state = OutputState.OUTPUTTING
            self._is_paused = False
            self._emit(OutputEventType.RESUME)

    def is_available(self) -> bool:
        """Check if adapter is available"""
        return True

    async def dispose(self) -> None:
        """Dispose resources"""
        self.stop()
        self._handlers.clear()
